ec/src/hashing/curve_maps/swu/mod.rs

```diff
@@ -161,10 +163,14 @@ impl<P: SWUParams> MapToCurve<GroupAffine<P>> for SWUMap<P> {
         let num_x = if gx1_square { num_x1 } else { num_x2 };
         let y = if gx1_square { y1 } else { y2 };
 
-        // 9. If sgn0(u) != sgn0(y), set y = -y
-        // let y = if y % 2 {-y} or {y};
+        
         let x_affine = num_x / div;
-        let y_affine = y;
+        // 9. If sgn0(u) != sgn0(y), set y = -y
+        let mut a = [0u8; 128];
+        let mut b =  [0u8; 128];
+        point.write(&mut a[..]).unwrap();
+        y.write(&mut b[..]).unwrap();
```

Comment:
@weikengchen @Pratyush I know this is inefficient, ugly and not suitable for production,  but we just left it there as a place holder of what needs to be done here, so we can ask you if  there is a better way of doing it. 

should we add a function like parity for ark_ff:field_trait and implement it for all models (as `self.c0.into_repr().is_odd()`  for example)?

Reply:
It is meant to support all curves. In particular, we need to support schemes where messages are hashed to G2 which is defined over Fp2.

Because the building of all extensions is precisely specified in both Arkworks and the Standard (the specific irreducible polynomial that should be used) one can agree on a specific basis {1,x,x^2,..} and define parity to be the parity of the coefficient of {1} basis element. 

Alternatively one can also use the parity of the norm, (I think trace is biased and is inconsistent across extensions) but this might be more expensive.

@daira has already solved this problem so we can dig into Zcash code to see their choice and implement it here. I was wondering if you think it is a good idea and if it hasn't been implemented before.

Reply:
I see. Let's add a method to the `Field` trait that returns an array of the underlying prime field elements then? So something like

```rust
pub trait Field {
	type BasePrimeFieldArr: AsRef<[Self::BasePrimeField]>
	fn to_base_prime_field_elements(&self) -> Self::BasePrimeFieldArr;
}
```

And then you can add a `parity` helper function here:

```rust
fn parity<F: Field>(element: &F) -> bool {
	element.to_base_prime_field_elements().as_ref()[0].into_repr().is_odd()
}
```

Reply:
In `pasta_curves` we use an `is_odd` method on field elements which is implemented (e.g. for the Pallas base field) [here](https://github.com/zcash/pasta_curves/blob/c052756831fd4721fcf6ebeb3f2bf83ee80a5cda/src/fields/fp.rs#L592-L608), which is equivalent to what @Pratyush suggests.

Reply:
@Pratyush could you tell me why do you prefer `fn parity<F: Field>(element: &F) -> bool` over simply adding a member method `fn parity(&self) -> bool` to Field trait?

Reply:
The reason is that `parity` isn't really a general operation that's defined for arbitrary fields, and is only really used in this hash-to-curve algorithm. So IMO it's not worth it to complicate the `Field` API with an additional method. On the other hand, adding an API of `BasePrimeFieldIter` makes sense, because you can always view a field element as a list of base prime field elements.